src/sync/mod.rs

```diff
@@ -139,6 +148,114 @@ pub enum BeginSyncError {
+        .map_err(NoBaseSnapshot)?;
+    let main_head_hash = dag_read
+        .get_head(db::DEFAULT_HEAD_NAME)
+        .await
+        .map_err(ReadError)?
+        .ok_or(MissingMainHead)?;
+    let main_snapshot = Commit::base_snapshot(&main_head_hash, &dag_read)
+        .await
+        .map_err(NoBaseSnapshot)?;
+    let meta = sync_snapshot.meta();
+    let sync_snapshot_basis = meta.basis_hash().ok_or(SyncSnapshotWithNoBasis)?;
+    if sync_snapshot_basis != main_snapshot.chunk().hash() {
+        return Err(OverlappingSyncs);
+    }
+
+    // Collect pending commits from the main chain and determine which
+    // of them if any need to be replayed.
+    let mut pending = Commit::pending(&main_head_hash, &dag_read)
+        .await
+        .map_err(PendingError)?;
+    let sync_head = Commit::from_hash(&sync_head_hash, &dag_read)
+        .await
+        .map_err(LoadCommitError)?;
+    pending.retain(|c| c.mutation_id() > sync_head.mutation_id());
+    // pending() gave us the pending mutations in sync-head-first order whereas
+    // caller wants them in the order to replay (lower mutation ids first).
+    pending.reverse();
+
+    // Return replay commits if any.
+    if !pending.is_empty() {
+        let mut replay_mutations: Vec<Mutation> = Vec::new();
+        for c in pending {
+            let (name, args) = match c.meta().typed() {
+                MetaTyped::Local(lm) => (
+                    lm.mutator_name().to_string(),
+                    lm.mutator_args_json().to_vec(),
+                ),
+                _ => return Err(ProgrammerError("pending mutation is not local".to_string())),
+            };
+            replay_mutations.push(Mutation {
+                id: c.mutation_id(),
+                name,
+                args,
+                original: Some(c.chunk().hash().to_string()),
+            })
+        }
+        return Ok(MaybeEndSyncResponse {
+            sync_head: sync_head_hash,
+            replay_mutations,
+        });
+    }
+
+    // TODO check invariants
+
+    // No mutations to replay so set the main head to the sync head and sync complete!
+    dag_write
+        .set_head(db::DEFAULT_HEAD_NAME, Some(&sync_head_hash))
+        .await
+        .map_err(WriteError)?;
+    dag_write
+        .set_head(SYNC_HEAD_NAME, None)
+        .await
+        .map_err(WriteError)?;
+    dag_write.commit().await.map_err(CommitError)?;
+    Ok(MaybeEndSyncResponse {
+        sync_head: sync_head_hash.to_string(),
+        replay_mutations: Vec::new(),
+    })
+}
+
+#[derive(Debug)]
+pub enum MaybeEndSyncError {
+    CommitError(dag::Error),
+    InvalidArgs(std::str::Utf8Error),
+    LoadCommitError(db::FromHashError),
+    MissingMainHead,
+    MissingSyncHead,
+    NoBaseSnapshot(db::BaseSnapshotError),
+    OverlappingSyncs,
+    PendingError(db::PendingError),
+    ProgrammerError(String),
+    ReadError(dag::Error),
+    SyncSnapshotWithNoBasis,
+    WriteError(dag::Error),
+    WrongSyncHead,
```

Comment:
OK there are admittedly a lot of cases here but this is maybe the most complex function in the entire system and I'd argue it's useful to understand the difference between all these different cases.